### ISSUE TITLE
The source code of Coq’s extraction plugin is using a lot of physical equalities

### ISSUE THREAD

**Issue by Mbodin** [coq/coq#12818]
#### Description of the problem

This is not really an issue with Coq itself, but with its source code. I’m currently looking at the extraction source code, and I’m noticing a lot of physical equalities (`==` and `!=`) against flags (that is, inductive with only a bunch of constructors with no argument).
See comment: https://github.com/coq/coq/commit/11ca3113365639136cf65a74c345080a9434e69b#r41369506
I don’t get it: using physical identity without a good reason is usually considered to be bad practice in OCaml. It makes a code already difficult to read even more difficult to read.
Is there a specific reason for using physical equality in this part of Coq?

**Comment by SkySkimmer** [comment]
> I don’t get it: using physical identity without a good reason is usually considered to be bad practice in OCaml. It makes a code already difficult to read even more difficult to read.

How so? Please explain more.

**Comment by Mbodin** [comment]
Sure. I find Coq’s source code quite difficult to dive in. Most functions are undocumented and use acronyms that are very difficult to get if one has never dived into a lot of Coq’s code before. In another word: its code documentation is not modular.
For instance, there is a type `env` defined in `common.mli`. It is clear that this is an environment, but is it not clear at all what kind of environment it is. Similarly, a newcomer will have great difficulty understanding what the pervasively used `mp` means and where to find it.

Don’t take me wrong: it’s quite manageable, and there are some documentation. But a lot is left for the user to guess by directly looking at the `ml` files instead of just the `mli` ones.

So I go to the `ml` files, and I see a lot of usage of the physical equality. Physical equality is not that used in the OCaml programs I’ve seen. So I’m thinking that there must be a reason: some hidden references and so on. But in the cases I suggest to change, there are no such things: they are just flags.

Also, the understand of most people of physical equality is the following: if `a == b`, then `a = b`, but the converse is not valid. Unfortunately, in the cases I suggest to move away from physical equality, the code assumes that the converse is valid: the code is invalid if we replace all calls to the physical equality by a constant function to `false`. This is quite bad: we are uselessly relying on a specific behaviour of the current compiler.

There are reasons for using physical equalities from time to time, like optimisation. For instance, my understanding of this `==` is that it prevents to reallocate a `MLlam` if nothing was changed: https://github.com/coq/coq/blob/master/plugins/extraction/mlutil.ml#L542 I’m fine with this particular usage as the code would still be valid if we were to replace `==` by `fun _ _ -> false` (the code would just be less efficient in memory).
However, in this place, replacing `==` by `fun _ _ -> false` would produce invalid output: https://github.com/coq/coq/blob/master/plugins/extraction/modutil.ml#L118 This kind of usage of `==` is not fine.

**Comment by Mbodin** [comment]
More precisely, the exact property that most people have in mind when thinking of physical equality is:
- `a == b` always immediately returns (no loops, no exception).
- if `a = b` does return, then `a == b` implies `a = b`.
In general `a = b` may loop if given recursive structure, or throw an exception if given functions.

This is what the documentation says about `==`: http://caml.inria.fr/pub/docs/manual-ocaml/libref/Stdlib.html
> On non-mutable types, the behavior of ( == ) is implementation-dependent; however, it is guaranteed that e1 == e2 implies compare e1 e2 = 0.

From this specification, a compiler that would hypothetically compile  `==` by `fun _ _ -> false` on any non-mutable structure (and in particular, flags), is valid. However, this would make the extraction as it is currently written invalid… which is bad.

**Comment by SkySkimmer** [comment]
In Coq we seek to avoid the polymorphic structural equality `(=)`:
- it's ill-behaved on some types (eg Names.KerName.t due to the lazily computed hash)
- some perf concerns (@ppedrot may know more about this)

(some uses remain from history or review oversight since we don't have a linter for it)

For complex types this means we define the equality function for the type (manually, because we haven't decided to add ppx stuff). 
For flags we are sometimes lazy and use physical equality `(==)`. A PR replacing instances of those by a manually defined equality function would probably be accepted AFAICT.

**Comment by Mbodin** [comment]
I understand these concerns. But I find the overuse of the physical equality in places where its specification is not enough to guarantee that things are all right quite problematical, especially given that trust is the very point of Coq.

Are these concerns documented? This is the first time that I’m told that `=` is not fine but `==` is on very simple types. ☹

**Comment by SkySkimmer** [comment]
BTW the comment on the `(==)` declaration used to say "On integers and characters, physical equality is identical to structural equality" but this was deleted in ocaml/ocaml@ed32f569e3b636e0f12efdbbd5bba9e05cc434ac "merge changes from ocaml3110 to ocaml3111rc0"

**Comment by Mbodin** [comment]
Pardon my indiscretion, but it is quite frustrating for me to read these concerns as-is: I fully agree that polymorphic equality has issues and its usage should be restricted (especially in opaque types)… but physical equality seems even worse given how it is loosely defined ☹ In particular, it doesn't solve at all the first concern, making it worse actually 😕​

Anyway, I was very surprised by how the code was written. Now I understand the development choices. Thanks for the feedback!

**Comment by ppedrot** [comment]
@Mbodin do you want to write an alternative to #12819 that relies on pattern-matching?

**Comment by Mbodin** [comment]
I agree that pattern-matching would be a better option. I unfortunately don’t have the time right now to do it properly: I don’t want to commit code into Coq in a hurry ☺ Maybe in a few weeks I will have more time. In the meantime, I’m fine with closing the PR.

I wouldn’t close the issue, though. I fully understand the concerns related to `=`, but `==` seems to actually be much worse: technically, `fun _ _ -> false` is a valid implementation of it… but it would break a lot of pieces of Coq as-is ☹

**Comment by ppedrot** [comment]
All right, let's close the PR and leave that issue open then.

**Comment by Zimmi48** [comment]
I agree with the expressed concerns and that it is not at all clear from the code why physical equality is used in some places, so a PR replacing those with pattern-matching would be very welcome.
